dali/imgcodec/tools/imagemagick_test.cc

```diff
@@ -0,0 +1,283 @@
+#include "dali/imgcodec/image_format.h"
+#include "dali/imgcodec/image_decoder.h"
+
+#include "dali/imgcodec/parsers/bmp.h"
+#include "dali/imgcodec/parsers/jpeg.h"
+#include "dali/imgcodec/parsers/jpeg2000.h"
+#include "dali/imgcodec/parsers/png.h"
+#include "dali/imgcodec/parsers/pnm.h"
+#include "dali/imgcodec/parsers/tiff.h"
+#include "dali/imgcodec/parsers/webp.h"
+
+
+const char *help = R"(Usage: imagemagick_test [OPTIONS] PATH
+
+Compares image shape returned by ImageMagick's `identify` tool with the shape
+returned by DALI's imgcodec. If PATH is a directory, it is searched recursively
+for matching files (see --filter). If PATH is a regular file, only this single
+file is processed and the filters do not apply.
+
+The following OPTIONS are available:
+--help   -h           Show this help message
+--filter=... -r=...   Regex to filter the files with, for example '.*\\.jpeg'
+--print  -p           Only print ImageMagick's output and don't compare it with
+--batch=N  -b=N       Runs ImageMagick on batches of N images (default is 1024)
+--identify=... -i=... Path of `identify` tool, default is /usr/bin/identify
+--jobs=N  -j=N        Number of concurrent jobs to run, defaults to number of cores
+)";
+
+
+namespace dali {
+namespace imgcodec {
+namespace test {
+
+void fail(std::ostream &stream, const std::string &filename, const std::string &message) {
+  stream << "FAIL" << "\t" << filename << ": " << message << std::endl;
+}
+
+void ok(std::ostream &stream, const std::string &filename) {
+  stream << "OK" << "\t" << filename << std::endl;
+}
+
+class ImgcodecTester {
+ public:
+  ImgcodecTester() {
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("jpeg", std::make_shared<JpegParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("png", std::make_shared<PngParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("bmp", std::make_shared<BmpParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("tiff", std::make_shared<TiffParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("pnm", std::make_shared<PnmParser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("jpeg2000", std::make_shared<Jpeg2000Parser>()));
+    format_registry_.RegisterFormat(
+        std::make_shared<ImageFormat>("webp", std::make_shared<WebpParser>()));
+  }
+
+  std::optional<TensorShape<>> shape_of(const std::string &filename) const {
+    auto img = ImageSource::FromFilename(filename);
+    auto fmt = this->format_registry_.GetImageFormat(&img);
+    if (fmt == nullptr) return {};
+    auto image_info = fmt->Parser()->GetInfo(&img);
+    return image_info.shape;
+  }
+
+  ImageFormatRegistry format_registry_;
+};
+
+struct Env {
+  std::string identify_path;
+  std::string directory;
+  std::regex filter;
+  bool print;
+  unsigned batch;
+  unsigned jobs;
```

Comment:
Please note that 3 fields above won't be initialized to any value.

Reply:
But is it a problem? They're assigned some defaults in `main` anyway